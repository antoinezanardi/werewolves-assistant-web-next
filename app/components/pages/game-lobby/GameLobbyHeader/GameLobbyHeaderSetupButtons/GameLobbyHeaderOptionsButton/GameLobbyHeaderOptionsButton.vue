<template>
  <PrimeVueButton
    id="game-lobby-header-options-button"
    icon="fa fa-sliders"
    :label="$t('components.GameLobbyHeaderOptionButton.gameOptions')"
    :pt="{ 'root': '!border-sky-600' }"
    severity="info"
    size="small"
    @click.prevent="onClickFromGameOptionsButton"
  />
</template>

<script setup lang="ts">
import type { GameLobbyHeaderOptionsButtonEmits } from "~/components/pages/game-lobby/GameLobbyHeader/GameLobbyHeaderSetupButtons/GameLobbyHeaderOptionsButton/game-lobby-header-options-button.types";

const emit = defineEmits<GameLobbyHeaderOptionsButtonEmits>();

function onClickFromGameOptionsButton(): void {
  emit("gameOptionsButtonClick");
}
</script>