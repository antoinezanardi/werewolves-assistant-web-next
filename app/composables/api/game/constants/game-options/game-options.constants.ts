import type { GameOptions } from "~/composables/api/game/types/game-options/game-options.class";

const DEFAULT_GAME_OPTIONS: GameOptions = {
  composition: { isHidden: false },
  votes: { canBeSkipped: true },
  roles: {
    areRevealedOnDeath: true,
    doSkipCallIfNoTarget: false,
    sheriff: {
      isEnabled: true,
      electedAt: {
        turn: 1,
        phaseName: "night",
      },
      hasDoubledVote: true,
      mustSettleTieInVotes: true,
    },
    bigBadWolf: { isPowerlessIfWerewolfDies: true },
    whiteWerewolf: { wakingUpInterval: 2 },
    seer: {
      isTalkative: true,
      canSeeRoles: true,
    },
    cupid: {
      lovers: { doRevealRoleToEachOther: false },
      mustWinWithLovers: false,
    },
    littleGirl: { isProtectedByDefender: false },
    defender: { canProtectTwice: false },
    elder: {
      livesCountAgainstWerewolves: 2,
      doesTakeHisRevenge: true,
    },
    idiot: { doesDieOnElderDeath: true },
    twoSisters: { wakingUpInterval: 2 },
    threeBrothers: { wakingUpInterval: 2 },
    fox: { isPowerlessIfMissesWerewolf: true },
    bearTamer: { doesGrowlOnWerewolvesSide: true },
    stutteringJudge: { voteRequestsCount: 1 },
    wildChild: { isTransformationRevealed: false },
    wolfHound: {
      isChosenSideRevealed: false,
      isSideRandomlyChosen: false,
    },
    thief: {
      mustChooseBetweenWerewolves: true,
      isChosenCardRevealed: false,
      additionalCardsCount: 2,
    },
    piedPiper: {
      charmedPeopleCountPerNight: 2,
      isPowerlessOnWerewolvesSide: true,
    },
    scandalmonger: { markPenalty: 2 },
    witch: { doesKnowWerewolvesTargets: true },
    prejudicedManipulator: { isPowerlessOnWerewolvesSide: true },
    actor: {
      isPowerlessOnWerewolvesSide: true,
      additionalCardsCount: 3,
    },
  },
} as const;

export { DEFAULT_GAME_OPTIONS };