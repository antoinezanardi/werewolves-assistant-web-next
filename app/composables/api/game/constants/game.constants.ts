import { PLAYER_GROUPS } from "~/composables/api/game/constants/player/player.constants";
import type { PlayerAttributeName } from "~/composables/api/game/types/players/player-attribute/player-attribute.types";
import type { PlayerGroup } from "~/composables/api/game/types/players/player.types";
import { ROLE_NAMES } from "~/composables/api/role/constants/role.constants";
import type { RoleName } from "~/composables/api/role/types/role.types";

const GAME_STATUSES = [
  "playing",
  "over",
  "canceled",
] as const;

const GAME_SOURCES = [
  ...PLAYER_GROUPS,
  ...Object.values(ROLE_NAMES),
  "sheriff",
] as const satisfies readonly (PlayerAttributeName | PlayerGroup | RoleName)[];

const MIN_PLAYERS_IN_GAME = 4;

const MAX_PLAYERS_IN_GAME = 40;

const MIN_PLAYERS_IN_GROUP = 2;

export {
  GAME_STATUSES,
  GAME_SOURCES,
  MIN_PLAYERS_IN_GAME,
  MAX_PLAYERS_IN_GAME,
  MIN_PLAYERS_IN_GROUP,
};