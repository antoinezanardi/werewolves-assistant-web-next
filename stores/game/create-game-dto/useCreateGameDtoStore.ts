import { defineStore } from "pinia";
import { ref } from "vue";

import { CreateGamePlayerDto } from "~/composables/api/game/dto/create-game/create-game-player/create-game-player.dto";
import { CreateGameDto } from "~/composables/api/game/dto/create-game/create-game.dto";
import { StoreIds } from "~/stores/enums/store.enum";

const useCreateGameDtoStore = defineStore(StoreIds.CREATE_GAME_DTO, () => {
  const createGameDto = ref<CreateGameDto>(CreateGameDto.create({ players: [] }));

  function setCreateGameDto(createGameDtoValue: CreateGameDto): void {
    createGameDto.value = CreateGameDto.create(createGameDtoValue);
  }

  function resetCreateGameDto(): void {
    createGameDto.value = CreateGameDto.create({ players: [] });
  }

  function addPlayerToCreateGameDto(player: CreateGamePlayerDto): void {
    createGameDto.value.players.push(CreateGamePlayerDto.create(player));
  }

  function updatePlayerInCreateGameDto(player: CreateGamePlayerDto): void {
    const playerIndex = createGameDto.value.players.findIndex(({ name }) => name === player.name);
    if (playerIndex !== -1) {
      createGameDto.value.players.splice(playerIndex, 1, CreateGamePlayerDto.create(player));
    }
  }

  function setPlayersToCreateGameDto(players: CreateGamePlayerDto[]): void {
    createGameDto.value.players = players.map(player => CreateGamePlayerDto.create(player));
  }

  function removePlayerFromCreateGameDto(playerName: string): void {
    const playerIndex = createGameDto.value.players.findIndex(player => player.name === playerName);
    if (playerIndex !== -1) {
      createGameDto.value.players = createGameDto.value.players.toSpliced(playerIndex, 1);
    }
  }
  return {
    createGameDto,
    setCreateGameDto,
    resetCreateGameDto,
    addPlayerToCreateGameDto,
    updatePlayerInCreateGameDto,
    setPlayersToCreateGameDto,
    removePlayerFromCreateGameDto,
  };
});

export { useCreateGameDtoStore };