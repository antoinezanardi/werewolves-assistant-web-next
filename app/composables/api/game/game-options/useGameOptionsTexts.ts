import { instanceToPlain } from "class-transformer";
import type { Paths } from "type-fest";
import type { ComputedRef, Ref } from "vue";
import { keys, get } from "radash";

import { DEFAULT_GAME_OPTIONS } from "~/composables/api/game/constants/game-options/game-options.constants";
import type { GameOptions } from "~/composables/api/game/types/game-options/game-options.class";
import type { DeepStringifiedGameOptions } from "~/composables/api/game/types/game-options/game-options.types";
import { useStrings } from "~/composables/misc/useStrings";
import { useTimers } from "~/composables/misc/useTimers";

type UseGameOptionsTexts = {
  gameOptionsTexts: ComputedRef<DeepStringifiedGameOptions>;
  changedGameOptionsTexts: ComputedRef<string[]>;
  getGameOptionText: (key: Paths<DeepStringifiedGameOptions>) => string;
  getGameOptionKeyFromText: (gameOptionText: string) => Paths<DeepStringifiedGameOptions> | undefined;
};

function useGameOptionsTexts(gameOptions: Ref<GameOptions>): UseGameOptionsTexts {
  const { convertBooleanAsAffirmativeString } = useStrings();
  const { getSecondsInMinutesLabel } = useTimers();
  const { t } = useI18n();

  const sheriffElectionTimeGameOptionText = computed<string>(() => {
    const phaseLabel = t(`shared.game.definitePhase.${gameOptions.value.roles.sheriff.electedAt.phaseName}`);

    return t("composables.useGameOptionsTexts.roles.sheriff.electedAt", {
      turn: gameOptions.value.roles.sheriff.electedAt.turn,
      phase: phaseLabel.toLowerCase(),
    });
  });

  const votesDurationGameOptionText = computed<string>(() => {
    const timeLabel = getSecondsInMinutesLabel(gameOptions.value.votes.duration);

    return t(`composables.useGameOptionsTexts.votes.duration`, { time: timeLabel });
  });

  const gameOptionsTexts = computed<DeepStringifiedGameOptions>(() => ({
    composition: { isHidden: t(`composables.useGameOptionsTexts.composition.isHidden.${convertBooleanAsAffirmativeString(gameOptions.value.composition.isHidden)}`) },
    votes: {
      canBeSkipped: t(`composables.useGameOptionsTexts.votes.canBeSkipped.${convertBooleanAsAffirmativeString(gameOptions.value.votes.canBeSkipped)}`),
      duration: votesDurationGameOptionText.value,
    },
    roles: {
      areRevealedOnDeath: t(`composables.useGameOptionsTexts.roles.areRevealedOnDeath.${convertBooleanAsAffirmativeString(gameOptions.value.roles.areRevealedOnDeath)}`),
      doSkipCallIfNoTarget: t(`composables.useGameOptionsTexts.roles.doSkipCallIfNoTarget.${convertBooleanAsAffirmativeString(gameOptions.value.roles.doSkipCallIfNoTarget)}`),
      werewolf: { canEatEachOther: t(`composables.useGameOptionsTexts.roles.werewolf.canEatEachOther.${convertBooleanAsAffirmativeString(gameOptions.value.roles.werewolf.canEatEachOther)}`) },
      sheriff: {
        isEnabled: t(`composables.useGameOptionsTexts.roles.sheriff.isEnabled.${convertBooleanAsAffirmativeString(gameOptions.value.roles.sheriff.isEnabled)}`),
        electedAt: {
          turn: sheriffElectionTimeGameOptionText.value,
          phaseName: sheriffElectionTimeGameOptionText.value,
        },
        hasDoubledVote: t(`composables.useGameOptionsTexts.roles.sheriff.hasDoubledVote.${convertBooleanAsAffirmativeString(gameOptions.value.roles.sheriff.hasDoubledVote)}`),
        mustSettleTieInVotes: t(`composables.useGameOptionsTexts.roles.sheriff.mustSettleTieInVotes.${convertBooleanAsAffirmativeString(gameOptions.value.roles.sheriff.mustSettleTieInVotes)}`),
      },
      bigBadWolf: { isPowerlessIfWerewolfDies: t(`composables.useGameOptionsTexts.roles.bigBadWolf.isPowerlessIfWerewolfDies.${convertBooleanAsAffirmativeString(gameOptions.value.roles.bigBadWolf.isPowerlessIfWerewolfDies)}`) },
      whiteWerewolf: { wakingUpInterval: t(`composables.useGameOptionsTexts.roles.whiteWerewolf.wakingUpInterval`, { interval: gameOptions.value.roles.whiteWerewolf.wakingUpInterval }, gameOptions.value.roles.whiteWerewolf.wakingUpInterval) },
      seer: {
        isTalkative: t(`composables.useGameOptionsTexts.roles.seer.isTalkative.${convertBooleanAsAffirmativeString(gameOptions.value.roles.seer.isTalkative)}`),
        canSeeRoles: t(`composables.useGameOptionsTexts.roles.seer.canSeeRoles.${convertBooleanAsAffirmativeString(gameOptions.value.roles.seer.canSeeRoles)}`),
      },
      cupid: {
        lovers: { doRevealRoleToEachOther: t(`composables.useGameOptionsTexts.roles.cupid.lovers.doRevealRoleToEachOther.${convertBooleanAsAffirmativeString(gameOptions.value.roles.cupid.lovers.doRevealRoleToEachOther)}`) },
        mustWinWithLovers: t(`composables.useGameOptionsTexts.roles.cupid.mustWinWithLovers.${convertBooleanAsAffirmativeString(gameOptions.value.roles.cupid.mustWinWithLovers)}`),
      },
      littleGirl: { isProtectedByDefender: t(`composables.useGameOptionsTexts.roles.littleGirl.isProtectedByDefender.${convertBooleanAsAffirmativeString(gameOptions.value.roles.littleGirl.isProtectedByDefender)}`) },
      defender: { canProtectTwice: t(`composables.useGameOptionsTexts.roles.defender.canProtectTwice.${convertBooleanAsAffirmativeString(gameOptions.value.roles.defender.canProtectTwice)}`) },
      elder: {
        livesCountAgainstWerewolves: t(`composables.useGameOptionsTexts.roles.elder.livesCountAgainstWerewolves`, { livesCount: gameOptions.value.roles.elder.livesCountAgainstWerewolves }, gameOptions.value.roles.elder.livesCountAgainstWerewolves),
        doesTakeHisRevenge: t(`composables.useGameOptionsTexts.roles.elder.doesTakeHisRevenge.${convertBooleanAsAffirmativeString(gameOptions.value.roles.elder.doesTakeHisRevenge)}`),
      },
      idiot: { doesDieOnElderDeath: t(`composables.useGameOptionsTexts.roles.idiot.doesDieOnElderDeath.${convertBooleanAsAffirmativeString(gameOptions.value.roles.idiot.doesDieOnElderDeath)}`) },
      twoSisters: { wakingUpInterval: t(`composables.useGameOptionsTexts.roles.twoSisters.wakingUpInterval`, { interval: gameOptions.value.roles.twoSisters.wakingUpInterval }, gameOptions.value.roles.twoSisters.wakingUpInterval) },
      threeBrothers: { wakingUpInterval: t(`composables.useGameOptionsTexts.roles.threeBrothers.wakingUpInterval`, { interval: gameOptions.value.roles.threeBrothers.wakingUpInterval }, gameOptions.value.roles.threeBrothers.wakingUpInterval) },
      fox: { isPowerlessIfMissesWerewolf: t(`composables.useGameOptionsTexts.roles.fox.isPowerlessIfMissesWerewolf.${convertBooleanAsAffirmativeString(gameOptions.value.roles.fox.isPowerlessIfMissesWerewolf)}`) },
      bearTamer: { doesGrowlOnWerewolvesSide: t(`composables.useGameOptionsTexts.roles.bearTamer.doesGrowlOnWerewolvesSide.${convertBooleanAsAffirmativeString(gameOptions.value.roles.bearTamer.doesGrowlOnWerewolvesSide)}`) },
      stutteringJudge: { voteRequestsCount: t(`composables.useGameOptionsTexts.roles.stutteringJudge.voteRequestsCount`, { voteRequestsCount: gameOptions.value.roles.stutteringJudge.voteRequestsCount }, gameOptions.value.roles.stutteringJudge.voteRequestsCount) },
      wildChild: { isTransformationRevealed: t(`composables.useGameOptionsTexts.roles.wildChild.isTransformationRevealed.${convertBooleanAsAffirmativeString(gameOptions.value.roles.wildChild.isTransformationRevealed)}`) },
      wolfHound: {
        isChosenSideRevealed: t(`composables.useGameOptionsTexts.roles.wolfHound.isChosenSideRevealed.${convertBooleanAsAffirmativeString(gameOptions.value.roles.wolfHound.isChosenSideRevealed)}`),
        isSideRandomlyChosen: t(`composables.useGameOptionsTexts.roles.wolfHound.isSideRandomlyChosen.${convertBooleanAsAffirmativeString(gameOptions.value.roles.wolfHound.isSideRandomlyChosen)}`),
      },
      thief: {
        mustChooseBetweenWerewolves: t(`composables.useGameOptionsTexts.roles.thief.mustChooseBetweenWerewolves.${convertBooleanAsAffirmativeString(gameOptions.value.roles.thief.mustChooseBetweenWerewolves)}`),
        isChosenCardRevealed: t(`composables.useGameOptionsTexts.roles.thief.isChosenCardRevealed.${convertBooleanAsAffirmativeString(gameOptions.value.roles.thief.isChosenCardRevealed)}`),
      },
      piedPiper: {
        charmedPeopleCountPerNight: t(`composables.useGameOptionsTexts.roles.piedPiper.charmedPeopleCountPerNight`, { charmedPeopleCountPerNight: gameOptions.value.roles.piedPiper.charmedPeopleCountPerNight }, gameOptions.value.roles.piedPiper.charmedPeopleCountPerNight),
        isPowerlessOnWerewolvesSide: t(`composables.useGameOptionsTexts.roles.piedPiper.isPowerlessOnWerewolvesSide.${convertBooleanAsAffirmativeString(gameOptions.value.roles.piedPiper.isPowerlessOnWerewolvesSide)}`),
        areCharmedPeopleRevealed: t(`composables.useGameOptionsTexts.roles.piedPiper.areCharmedPeopleRevealed.${convertBooleanAsAffirmativeString(gameOptions.value.roles.piedPiper.areCharmedPeopleRevealed)}`),
      },
      scandalmonger: { markPenalty: t(`composables.useGameOptionsTexts.roles.scandalmonger.markPenalty`, { markPenalty: gameOptions.value.roles.scandalmonger.markPenalty }, gameOptions.value.roles.scandalmonger.markPenalty) },
      witch: { doesKnowWerewolvesTargets: t(`composables.useGameOptionsTexts.roles.witch.doesKnowWerewolvesTargets.${convertBooleanAsAffirmativeString(gameOptions.value.roles.witch.doesKnowWerewolvesTargets)}`) },
      prejudicedManipulator: { isPowerlessOnWerewolvesSide: t(`composables.useGameOptionsTexts.roles.prejudicedManipulator.isPowerlessOnWerewolvesSide.${convertBooleanAsAffirmativeString(gameOptions.value.roles.prejudicedManipulator.isPowerlessOnWerewolvesSide)}`) },
      actor: {
        isPowerlessOnWerewolvesSide: t(`composables.useGameOptionsTexts.roles.actor.isPowerlessOnWerewolvesSide.${convertBooleanAsAffirmativeString(gameOptions.value.roles.actor.isPowerlessOnWerewolvesSide)}`),
      },
    },
  }));

  const changedGameOptionsTexts = computed<string[]>(() => {
    const gameOptionsKeys = keys(instanceToPlain(gameOptions.value));
    const changedGameOptionsKeys = gameOptionsKeys.filter(key => {
      const value = get(gameOptions.value, key);
      const defaultValue = get(DEFAULT_GAME_OPTIONS, key);

      return value !== defaultValue;
    });

    return [...new Set(changedGameOptionsKeys.map(key => getGameOptionText(key as Paths<DeepStringifiedGameOptions>)))];
  });

  function getGameOptionText(key: Paths<DeepStringifiedGameOptions>): string {
    return get(gameOptionsTexts.value, key as string);
  }

  function getGameOptionKeyFromText(gameOptionText: string): Paths<DeepStringifiedGameOptions> | undefined {
    const gameOptionsTextsKeys = keys(gameOptionsTexts.value);
    const gameOptionKey = gameOptionsTextsKeys.find(key => get(gameOptionsTexts.value, key) === gameOptionText);

    return gameOptionKey as Paths<DeepStringifiedGameOptions>;
  }
  return {
    gameOptionsTexts,
    changedGameOptionsTexts,
    getGameOptionText,
    getGameOptionKeyFromText,
  };
}

export { useGameOptionsTexts };