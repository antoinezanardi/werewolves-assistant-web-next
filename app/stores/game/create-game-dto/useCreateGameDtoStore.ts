import { defineStore } from "pinia";
import { ref } from "vue";
import { DEFAULT_GAME_OPTIONS } from "~/composables/api/game/constants/game-options/game-options.constants";
import type { CreateGameAdditionalCardDto } from "~/composables/api/game/dto/create-game/create-game-additional-card/create-game-additional-card.dto";

import { CreateGamePlayerDto } from "~/composables/api/game/dto/create-game/create-game-player/create-game-player.dto";
import { CreateGameDto } from "~/composables/api/game/dto/create-game/create-game.dto";
import type { GameAdditionalCardRecipientRoleName } from "~/composables/api/game/types/game-additional-card/game-additional-card.types";
import { GameOptions } from "~/composables/api/game/types/game-options/game-options.class";
import { ADDITIONAL_CARDS_DEPENDANT_ROLES } from "~/composables/api/role/constants/role.constants";
import type { RoleName } from "~/composables/api/role/types/role.types";
import { StoreIds } from "~/stores/enums/store.enum";
import { useRolesStore } from "~/stores/role/useRolesStore";
import { useLocalStorage } from "@vueuse/core";
import { LocalStorageKeys } from "~/utils/enums/local-storage.enums";

const useCreateGameDtoStore = defineStore(StoreIds.CREATE_GAME_DTO, () => {
  const rolesStore = useRolesStore();
  const { getRoleWithNameInRoles } = rolesStore;

  const createGameOptionsDtoFromLocalStorage = useLocalStorage(LocalStorageKeys.GAME_OPTIONS, DEFAULT_GAME_OPTIONS, { mergeDefaults: true });

  const defaultCreateGameDto = CreateGameDto.create({
    players: [],
    options: DEFAULT_GAME_OPTIONS,
  });

  const createGameDto = ref<CreateGameDto>(CreateGameDto.create({
    ...defaultCreateGameDto,
    options: createGameOptionsDtoFromLocalStorage.value,
  }));

  const createGameOptionsDto = computed<GameOptions>(() => createGameDto.value.options);

  const doesCreateGameDtoContainPositionDependantRoles = computed<boolean>(() => {
    const positionDependantRoles: RoleName[] = ["rusty-sword-knight", "bear-tamer", "fox"];
    const playersWithPositionDependantRoles = getPlayersWithAnyRoleNameInCreateGameDto(positionDependantRoles);

    return playersWithPositionDependantRoles.length > 0;
  });

  const doesCreateGameDtoContainAdditionalCardsDependantRoles = computed<boolean>(() => {
    const playersWithAdditionalCardsDependantRoles = getPlayersWithAnyRoleNameInCreateGameDto([...ADDITIONAL_CARDS_DEPENDANT_ROLES]);

    return playersWithAdditionalCardsDependantRoles.length > 0;
  });

  function setCreateGameDto(createGameDtoValue: CreateGameDto): void {
    createGameDto.value = CreateGameDto.create(createGameDtoValue);
    saveCreateGameOptionsDtoToLocalStorage();
  }

  function resetCreateGameDto(doesRetrieveLocalStorageValues = true): void {
    if (!doesRetrieveLocalStorageValues) {
      createGameDto.value = CreateGameDto.create(defaultCreateGameDto);
      saveCreateGameOptionsDtoToLocalStorage();

      return;
    }
    createGameDto.value = CreateGameDto.create({
      ...defaultCreateGameDto,
      options: createGameOptionsDtoFromLocalStorage.value,
    });
  }

  function resetCreateGameOptionsDto(): void {
    createGameDto.value.options = GameOptions.create(DEFAULT_GAME_OPTIONS);
    saveCreateGameOptionsDtoToLocalStorage();
  }

  function saveCreateGameOptionsDtoToLocalStorage(): void {
    createGameOptionsDtoFromLocalStorage.value = createGameDto.value.options;
  }

  function removeObsoleteAdditionalCardsFromCreateGameDto(): void {
    if (!createGameDto.value.additionalCards) {
      return;
    }
    const rolesInGame = new Set(createGameDto.value.players.map(player => player.role.name));
    createGameDto.value.additionalCards = createGameDto.value.additionalCards.filter(({ recipient }) => rolesInGame.has(recipient));
  }

  function addPlayerToCreateGameDto(player: CreateGamePlayerDto): void {
    createGameDto.value.players.push(CreateGamePlayerDto.create(player));
  }

  function updatePlayerInCreateGameDto(player: CreateGamePlayerDto): void {
    const playerIndex = createGameDto.value.players.findIndex(({ name }) => name === player.name);
    if (playerIndex !== -1) {
      createGameDto.value.players.splice(playerIndex, 1, CreateGamePlayerDto.create(player));
    }
  }

  function setPlayersToCreateGameDto(players: CreateGamePlayerDto[]): void {
    createGameDto.value.players = players.map(player => CreateGamePlayerDto.create(player));
  }

  function removePlayerFromCreateGameDto(playerName: string): void {
    const playerIndex = createGameDto.value.players.findIndex(player => player.name === playerName);
    if (playerIndex !== -1) {
      createGameDto.value.players = createGameDto.value.players.toSpliced(playerIndex, 1);
    }
  }

  function isRoleInCreateGameDto(roleName: RoleName): boolean {
    return createGameDto.value.players.some(player => player.role.name === roleName);
  }

  function getPlayersWithRoleNameInCreateGameDto(roleName: RoleName): CreateGamePlayerDto[] {
    return createGameDto.value.players.filter(player => player.role.name === roleName);
  }

  function getPlayersWithAnyRoleNameInCreateGameDto(roleNames: RoleName[]): CreateGamePlayerDto[] {
    return createGameDto.value.players.filter(player => player.role.name && roleNames.includes(player.role.name));
  }

  function isRoleMinReachedInCreateGameDto(roleName: RoleName): boolean {
    const playersWithRoleName = getPlayersWithRoleNameInCreateGameDto(roleName);
    const role = getRoleWithNameInRoles(roleName);

    return !!role && (role.minInGame === undefined || playersWithRoleName.length >= role.minInGame);
  }

  function isRoleMaxReachedInCreateGameDto(roleName: RoleName): boolean {
    const playersWithRoleName = getPlayersWithRoleNameInCreateGameDto(roleName);
    const role = getRoleWithNameInRoles(roleName);

    return !!role && playersWithRoleName.length >= role.maxInGame;
  }

  function getRoleLeftCountToReachMinInCreateGameDto(roleName: RoleName): number {
    const playersWithRoleName = getPlayersWithRoleNameInCreateGameDto(roleName);
    const role = getRoleWithNameInRoles(roleName);
    if (role?.minInGame === undefined) {
      return 0;
    }
    const leftCount = role.minInGame - playersWithRoleName.length;

    return leftCount <= -1 ? 0 : leftCount;
  }

  function setAdditionalCardsForRecipientInCreateGameDto(additionalCards: CreateGameAdditionalCardDto[], recipient: GameAdditionalCardRecipientRoleName): void {
    if (!createGameDto.value.additionalCards) {
      createGameDto.value.additionalCards = additionalCards;
    }

    createGameDto.value.additionalCards = createGameDto.value.additionalCards.filter(card => card.recipient !== recipient);
    createGameDto.value.additionalCards.push(...additionalCards);
  }

  function getAdditionalCardsForRecipientInCreateGameDto(recipient: GameAdditionalCardRecipientRoleName): CreateGameAdditionalCardDto[] {
    return createGameDto.value.additionalCards?.filter(card => card.recipient === recipient) ?? [];
  }

  function getAdditionalCardsWithRoleNameInCreateGameDto(roleName: RoleName): CreateGameAdditionalCardDto[] {
    return createGameDto.value.additionalCards?.filter(card => card.roleName === roleName) ?? [];
  }
  return {
    createGameOptionsDtoFromLocalStorage,
    createGameDto,
    createGameOptionsDto,
    doesCreateGameDtoContainPositionDependantRoles,
    doesCreateGameDtoContainAdditionalCardsDependantRoles,
    setCreateGameDto,
    resetCreateGameDto,
    resetCreateGameOptionsDto,
    saveCreateGameOptionsDtoToLocalStorage,
    removeObsoleteAdditionalCardsFromCreateGameDto,
    addPlayerToCreateGameDto,
    updatePlayerInCreateGameDto,
    setPlayersToCreateGameDto,
    removePlayerFromCreateGameDto,
    isRoleInCreateGameDto,
    getPlayersWithRoleNameInCreateGameDto,
    getPlayersWithAnyRoleNameInCreateGameDto,
    isRoleMinReachedInCreateGameDto,
    isRoleMaxReachedInCreateGameDto,
    getRoleLeftCountToReachMinInCreateGameDto,
    setAdditionalCardsForRecipientInCreateGameDto,
    getAdditionalCardsForRecipientInCreateGameDto,
    getAdditionalCardsWithRoleNameInCreateGameDto,
  };
});

export { useCreateGameDtoStore };