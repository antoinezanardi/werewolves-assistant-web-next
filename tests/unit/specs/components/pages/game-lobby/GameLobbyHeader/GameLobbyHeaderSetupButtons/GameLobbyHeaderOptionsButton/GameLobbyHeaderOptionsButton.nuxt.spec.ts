import { createTestingPinia } from "@pinia/testing";
import { createFakeCreateGameDto } from "@tests/unit/utils/factories/composables/api/game/dto/create-game/create-game.dto.factory";
import { createFakeCompositionGameOptions } from "@tests/unit/utils/factories/composables/api/game/game-options/composition-game-options/composition-game-options.factory";
import { createFakeGameOptions } from "@tests/unit/utils/factories/composables/api/game/game-options/game-options.factory";
import { pTooltipDirectiveBinder } from "@tests/unit/utils/helpers/directive.helpers";
import type { BoundTooltip } from "@tests/unit/utils/types/directive.types";
import type { mount } from "@vue/test-utils";
import type { ComponentMountingOptions } from "@vue/test-utils/dist/mount";
import type Badge from "primevue/badge";
import type Button from "primevue/button";
import GameLobbyHeaderOptionsButton from "~/components/pages/game-lobby/GameLobbyHeader/GameLobbyHeaderSetupButtons/GameLobbyHeaderOptionsButton/GameLobbyHeaderOptionsButton.vue";

import { mountSuspendedComponent } from "@tests/unit/utils/helpers/mount.helpers";
import { DEFAULT_GAME_OPTIONS } from "~/composables/api/game/constants/game-options/game-options.constants";
import { StoreIds } from "~/stores/enums/store.enum";
import { useCreateGameDtoStore } from "~/stores/game/create-game-dto/useCreateGameDtoStore";

describe("Game Lobby Header Options Button Component", () => {
  const defaultCreateGameDto = createFakeCreateGameDto({
    options: DEFAULT_GAME_OPTIONS,
  });
  const testingPinia = { initialState: { [StoreIds.CREATE_GAME_DTO]: { createGameDto: defaultCreateGameDto } } };
  let wrapper: ReturnType<typeof mount<typeof GameLobbyHeaderOptionsButton>>;

  async function mountGameLobbyHeaderOptionsButtonComponent(options: ComponentMountingOptions<typeof GameLobbyHeaderOptionsButton> = {}):
  Promise<ReturnType<typeof mount<typeof GameLobbyHeaderOptionsButton>>> {
    return mountSuspendedComponent(GameLobbyHeaderOptionsButton, {
      global: {
        plugins: [createTestingPinia(testingPinia)],
        stubs: {
          Button: false,
        },
      },
      ...options,
    });
  }

  beforeEach(async() => {
    wrapper = await mountGameLobbyHeaderOptionsButtonComponent();
    const createGameDtoStore = useCreateGameDtoStore();
    createGameDtoStore.createGameDto = createFakeCreateGameDto(defaultCreateGameDto);
  });

  it("should match snapshot when rendered.", () => {
    expect(wrapper).toBeTruthy();
    expect(wrapper.html()).toMatchSnapshot();
  });

  describe("Options Button", () => {
    it("should translate button label when rendered.", () => {
      const button = wrapper.findComponent<typeof Button>("#game-lobby-header-options-button");

      expect(button.text()).toBe("Game options");
    });

    it("should render changed option badge when game options have been changed.", async() => {
      const createGameDtoStore = useCreateGameDtoStore();
      createGameDtoStore.createGameDto.options.composition.isHidden = true;
      await nextTick();
      const badge = wrapper.findComponent<typeof Badge>("#changed-game-options-count-badge");

      expect(badge.attributes("value")).toBe("1");
    });

    it("should not render changed option badge when game options have not been changed.", () => {
      const badge = wrapper.findComponent<typeof Badge>("#changed-game-options-count-badge");

      expect(badge.exists()).toBeFalsy();
    });

    it("should attach tooltip to badge when rendered.", async() => {
      const tooltip: BoundTooltip = { value: undefined };
      const directives = { ...pTooltipDirectiveBinder(tooltip, "#changed-game-options-count-badge") };
      wrapper = await mountGameLobbyHeaderOptionsButtonComponent({
        global: {
          directives,
          plugins: [
            createTestingPinia({
              initialState: {
                [StoreIds.CREATE_GAME_DTO]: {
                  createGameDto: createFakeCreateGameDto({
                    options: createFakeGameOptions({
                      ...DEFAULT_GAME_OPTIONS,
                      composition: createFakeCompositionGameOptions({ isHidden: true }),
                    }),
                  }),
                },
              },
            }),
          ],
          stubs: {
            Button: false,
          },
        },
      });

      expect(tooltip.value).toBe(`components.GameLobbyHeaderOptionsButton.youHaveChangedGameOptions, {"count":1}`);
    });

    describe("Click on button", () => {
      it("should emit 'gameOptionsButtonClick' event when clicked.", async() => {
        const button = wrapper.findComponent<typeof Button>("#game-lobby-header-options-button");
        await button.trigger("click");

        expect(wrapper.emitted("gameOptionsButtonClick")).toBeTruthy();
      });
    });
  });
});