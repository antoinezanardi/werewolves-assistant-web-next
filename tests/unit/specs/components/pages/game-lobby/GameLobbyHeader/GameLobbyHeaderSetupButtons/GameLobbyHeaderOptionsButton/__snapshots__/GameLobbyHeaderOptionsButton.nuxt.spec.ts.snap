// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Game Lobby Header Options Button Component > should match snapshot when rendered. 1`] = `
"<button class="p-button p-component p-button-info p-button-sm !border-sky-600" type="button" data-pc-name="button" data-p-disabled="false" data-p-severity="info" id="game-lobby-header-options-button" pc1="" data-pc-section="root" data-pd-ripple="true" style="overflow: hidden; position: relative;">
  <font-awesome-icon-stub icon="magic-wand-sparkles" border="false" fixedwidth="false" flip="false" listitem="false" pulse="false" swapopacity="false" spin="false" symbol="false" inverse="false" bounce="false" shake="false" beat="false" fade="false" beatfade="false" flash="false" spinpulse="false" spinreverse="false"></font-awesome-icon-stub><span id="button-text" class="hidden md:inline">Game options</span>
  <!--v-if--><span role="presentation" aria-hidden="true" data-p-ink="true" data-p-ink-active="false" class="p-ink" pd1="" data-pc-name="ripple" data-pc-section="root"></span>
</button>"
`;
